drivers/acpi/scan.c

```diff
@@ -2114,30 +2108,58 @@ static void acpi_bus_attach(struct acpi_device *device, bool first_pass)
 		device->handler->hotplug.notify_online(device);
 }
 
-void acpi_walk_dep_device_list(acpi_handle handle)
+static int __acpi_dev_flag_dependency_met(struct acpi_dep_data *dep,
+					  void *data)
 {
-	struct acpi_dep_data *dep, *tmp;
 	struct acpi_device *adev;
 
+	acpi_bus_get_device(dep->consumer, &adev);
+	if (!adev)
+		return 0;
+
+	adev->dep_unmet--;
+	if (!adev->dep_unmet)
+		acpi_bus_attach(adev, true);
+
+	list_del(&dep->node);
+	kfree(dep);
+	return 0;
+}
+
+void acpi_walk_dep_device_list(acpi_handle handle,
+			       int (*callback)(struct acpi_dep_data *, void *),
+			       void *data)
+{
+	struct acpi_dep_data *dep, *tmp;
+	int ret;
+
 	mutex_lock(&acpi_dep_list_lock);
 	list_for_each_entry_safe(dep, tmp, &acpi_dep_list, node) {
 		if (dep->supplier == handle) {
-			acpi_bus_get_device(dep->consumer, &adev);
-
-			if (adev) {
-				adev->dep_unmet--;
-				if (!adev->dep_unmet)
-					acpi_bus_attach(adev, true);
-			}
-
-			list_del(&dep->node);
-			kfree(dep);
+			ret = callback(dep, data);
```

Comment:
Hm, I think positive / negative values might be slightly better on the grounds that depending on the callback it might _always_ result in a break.  I'll have a play with that, thanks.

Reply:
I think positive/negative makes more sense given the context, i.e. the other functions there also return `int` and not `acpi_status`.

Reply:
OK; I documented that behaviour and added the return values. I've left `acpi_dev_flag_dependency_met()` a void, but I think this was a good change, thanks.